### ISSUE TITLE
Proposal: get language list from Crowdin API

### ISSUE THREAD

**Issue by ImRodry** [crowdin/ota-client-js#32]
Recently with #27 I realized that, by hardcoding the response from the Crowdin languages API endpoint, this might quickly become outdated and need an update through a PR. In order to fix that, I've thought of two possible solutions:

- Have the languages in a json file that gets updated daily through a GitHub action
- Add a language cache that stores the languages after they're needed and is null on startup. Since this endpoint doesn't require an API key, we can just make an extra API call once and it should be accurate throughout the rest of the session, since this list doesn't update often and projects also take a long time to add new languages.

Please let me know which one you'd prefer and I'd be willing to open a PR about this, although the solutions I've been experimenting with require a big refactor.

**Comment by andrii-bodnar** [comment]
Hi @ImRodry!

Thanks a lot for your suggestion. Discussed it with the team and from our perspective, the best way will be to make an API call and fetch the languages list. 

You can use this API method - https://support.crowdin.com/api/v2/#operation/api.languages.getMany

Also, we should yo think about some caching for the response to reduce the number of API calls (languages are not changed so often). Maybe, you will have some ideas?

**Comment by ImRodry** [comment]
> Also, we should think about some caching for the response to reduce the number of API calls (languages are not changed so often). Maybe, you will have some ideas?

Yeah I was thinking of using that endpoint but only after the language list is needed, as we would not be able to fetch it upon instantiating the class. This way you could have a languagesCache private prop and only fetch if that prop is null.
My issue with this and the reason I didn't make a PR to begin with is that this would require pretty much all functions in exportPattern.ts to be moved to the class so that they could access the property, which I would like to know if you think is fine to do or if there's another alternative.

**Comment by yevheniyJ** [comment]
We do have already some kind of caching mechanism for manifest and I think we can implmenet it in the same way. Check the `manifest` getter method. Regarding the changes in `exportPattern.ts` I think what we can do is to simply extend the each method to accept the list of languages. Therefore we can keep all functions in the current place. What is still not yet clear is how to get languages for Enterprise usage. The endpoint is then dynamic and organization is a part of it. And with having only `hash` it is not possible to determine what url needs to be called for list of languages. @andrii-bodnar  Do you think we can still use v2 API endpoint for Enterprise usecases as well?

**Comment by ImRodry** [comment]
It seems like the Enterprise API needs the name of the organization, however, it does not need an API token. The only difference from the regular crowdin endpoint to the enterprise one is that some languages are missing and the locales for lolcat and pirate english are different. Perhaps they could be updated to match? Either way I believe it'd be safer to stick with the regular API as we don't need the org name and it is more complete, however we can also add an option to the client for the organization name which, when provided, will fetch the crowdin enterprise API

**Comment by yevheniyJ** [comment]
That might be option but the only concern that I have here is that users can configure client incorrectly (e.g. specifying organization but they are actually not using the enterprise version). We at least should write it down in the documentation.

**Comment by ImRodry** [comment]
I'll make sure to include proper documentation as well as a note in the README when I make the PR 👍🏻

**Comment by andrii-bodnar** [comment]
@yevheniyJ @ImRodry we can start with using only the regular Crowdin API and then implement Crowdin Enterprise API support. The difference in languages is small and in most cases, it will work well. 

We are considering the possibility to include the organization name into the distribution `manifest.json` and then it can be fetched to make a corresponding API request using the Crowdin Enterprise API endpoint.

Anyway, from my perspective, I don't see anything wrong in manually passing the organization name to OTA Client except a case when it can be configured incorrectly as @yevheniyJ said.
